### PR TITLE
RawDataRecord with unit tests

### PR DESCRIPTION
**Pull request by n135c10r**
# Description
Draft of Implementation for RawDataRecord with unit tests.
Refactoring the Data Records directory structure.

# How Has This Been Tested?
Executing unit tests.

# Process

I know the [process](https://github.com/mdabrowski1990/uds/wiki/Software-Development-Life-Cycle) and did my best to follow it

### CODE REVIEW

File: tests/software_tests/database/data_record/test_abstract_data_record.py [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import pytest
 from mock import Mock, patch
 
-from uds.database.abstract_data_record import AbstractDataRecord
+from uds.database.data_record import AbstractDataRecord
```
Comment (mdabrowski1990) [inline tests/software_tests/database/data_record/test_abstract_data_record.py:4]: Let's import directly from the module `uds.database.data_record.abstract_data_record`

File: uds/database/__init__.py [hunk 2]
```diff
@@ -4,4 +4,4 @@
 Tools for decoding and encoding information from/to diagnostic messages.
 """
 
-from .abstract_data_record import AbstractDataRecord, DataRecordType, DecodedDataRecord
+from .data_record import RawDataRecord
```
Comment (mdabrowski1990) [inline uds/database/__init__.py:7]: we need to provide `AbstractDataRecord` and `DecodedDataRecord` as well

File: uds/database/data_record/abstract_data_record.py
Comment (mdabrowski1990) [inline uds/database/data_record/abstract_data_record.py:1]: After thinking it through, I am pretty sure we do not need `DataRecordType` class at all (we might use classes). You might remove it or keep it as it is and I would remove it at the first occasion.

File: uds/database/data_record/raw_data_record.py
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:1]: We need module docstring.

File: uds/database/data_record/raw_data_record.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        Initialization of Raw Data Record.
+
+        :param name: Name to assign to this Data Record.
+        :param length: Number of bits that this Raw Data Record is stored over.
+
+        :raise TypeError: Provided name is not str type.
+        :raise ValueError: Provided length is not a positive integer.
+        """
+        super().__init__(name)
+        if not isinstance(length, int) or length <= 0:
```
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:20]: I would do: `self.length = length`  here. Instead provide setter for `length` property. There you can handle type and value errors. Then you would handle all improper types and values assignment, not just on object creation.

You can base on https://github.com/mdabrowski1990/uds/blob/main/uds/transport_interface/can/common.py (e.g. `n_as_timeout` property)

File: uds/database/data_record/raw_data_record.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    def decode(self, raw_value: int) -> DecodedDataRecord:
+        """
+        Decode physical value for provided raw value.
+
+        :param raw_value: Raw (bit) value of Data Record.
+
+        :return: Dictionary with physical value for this Data Record.
+        """
+        return DecodedDataRecord(name=self.name, raw_value=raw_value, physical_value="Raw Data Record")
```
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:79]: It should be `DecodedDataRecord(name=self.name, raw_value=raw_value, physical_value=raw_value)` (this is an edge case, cause Raw Data Record do not have mapping, so physical and raw value are the same.

File: tests/software_tests/database/data_record/test_raw_data_record.py
Comment (mdabrowski1990) [inline tests/software_tests/database/data_record/test_raw_data_record.py:1]: I would appreciate, if you could extend these tests (I know that I make very comprehensive and deep testing, so just let me know if you think this is too much for you and I will finish this).

I would like to cover with tests:
- multiple cases to make sure that `__init__`, `decode` and `encode` methods do not malfunction with various parameters
- we do 100% branch and decisions coverage (so every error and code path has to be covered)

You can base on: https://github.com/mdabrowski1990/uds/blob/main/tests/software_tests/transport_interface/can/test_common.py

**Review by mdabrowski1990 — CHANGES_REQUESTED** [review]
I like that you separated `data_record` module.
I have left plenty of comments, but I think this is a really good start.

File: tests/software_tests/database/data_record/test_raw_data_record.py
Comment (n135c10r) [reply tests/software_tests/database/data_record/test_raw_data_record.py:1]: Ok, no problem. I will do it tomorrow. Please check if rest of the fixes are acceptable :)

File: uds/database/data_record/__init__.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+"""
+Package with implementation for all type of Data Records.
+
+Each Data Record contains mapping (translation) of raw data (sequence of bits in diagnostic message payload) to some
+meaningful information (e.g. physical value, text).
+"""
+
+__all__ = ["AbstractDataRecord", "DecodedDataRecord", "RawDataRecord"]
```
Comment (mdabrowski1990) [inline uds/database/data_record/__init__.py:8]: please remove `__all__` (everything we import is supposed to be available for the user)

File: tests/software_tests/database/data_record/test_abstract_data_record.py [hunk 6]
```diff
@@ -35,3 +35,7 @@ def test_init__valid(self, mock_isinstance):
     def test_name(self):
         self.mock_data_record._AbstractDataRecord__name = Mock()
         assert AbstractDataRecord.name.fget(self.mock_data_record) == self.mock_data_record._AbstractDataRecord__name
+
+    def test_data_record_type(self):
```
Comment (mdabrowski1990) [inline tests/software_tests/database/data_record/test_abstract_data_record.py:39]: let's remove this

File: uds/database/data_record/raw_data_record.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    def decode(self, raw_value: int) -> DecodedDataRecord:
+        """
+        Decode physical value for provided raw value.
+
+        :param raw_value: Raw (bit) value of Data Record.
+
+        :return: Dictionary with physical value for this Data Record.
+        """
+        return DecodedDataRecord(name=self.name, raw_value=raw_value, physical_value=raw_value)
```
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:92]: let's protect from:
- user providing value of other than int type -> Type Error
- user providing value out of range -> Value Error

tip: To make check whether value is in range, you might want to add property `max_raw_value` (equals `(1<<length)-1`)

File: uds/database/data_record/raw_data_record.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    def encode(self, physical_value: DataRecordPhysicalValueAlias) -> int:
+        """
+        Encode raw value for provided physical value.
+
+        :param physical_value: Physical (meaningful e.g. float, str type) value of this Data Record.
+
+        :return: Raw Value of this Data Record.
+        """
+        return physical_value  # type: ignore
```
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:102]: lets protect from wrong type and value out of range provided by the user

File: uds/database/data_record/raw_data_record.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        :raise ValueError: Provided value is less or equal 0.
+        """
+        if not isinstance(value, int):
+            raise TypeError("Length must be an integer.")
+        if value <= 0:
+            raise ValueError("Length must be a positive integer.")
+        self.__length = value
+
+    @property
+    def max_raw_value(self):
```
Comment (mdabrowski1990) [inline uds/database/data_record/raw_data_record.py:48]: let's move it to `AbstractDataRecord` (it is common for all DataRecords)

File: tests/software_tests/database/data_record/test_raw_data_record.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            raw_value=value,
+            physical_value=value
+        )
+
+    @pytest.mark.parametrize(
+        "value", ["test", None]
+    )
+    @patch(f"{SCRIPT_LOCATION}.isinstance")
+    def test_decode_type_error(self, mock_isinstance, value):
+        with pytest.raises(TypeError):
```
Comment (mdabrowski1990) [inline tests/software_tests/database/data_record/test_raw_data_record.py:111]: we need one more before:
`mock_isinstance.return_value = False` to make sure that type check failed

File: tests/software_tests/database/data_record/test_raw_data_record.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    def test_encode(self, value):
+        self.mock_data_record.max_raw_value = 8
+        assert RawDataRecord.encode(self.mock_data_record, value) == value
+
+    @pytest.mark.parametrize(
+        "value", ["test", None]
+    )
+    @patch(f"{SCRIPT_LOCATION}.isinstance")
+    def test_encode_type_error(self, mock_isinstance, value):
+        with pytest.raises(TypeError):
```
Comment (mdabrowski1990) [inline tests/software_tests/database/data_record/test_raw_data_record.py:141]: we need one more line before with pytest.raises:
mock_isinstance.return_value = False to make sure that type check failed

**Review by mdabrowski1990 — COMMENTED** [review]
almost ready to be merged
please check why codecov reports missing coverage (there is something wrong with TypeError tests, probably it would be solved after my comments are applied)

**Review by mdabrowski1990 — APPROVED** [review]
